File: Scripts/V3 24-06-2025/getMacros.php
<?php

header("Access-Control-Allow-Origin: *");
header("Access-Control-Allow-Methods: GET, OPTIONS");
header("Access-Control-Allow-Headers: Content-Type, Authorization, X-API-KEY");
header('Content-Type: application/json');

date_default_timezone_set('UTC');

// Handle CORS preflight request
if ($_SERVER['REQUEST_METHOD'] === 'OPTIONS') {
    http_response_code(200);
    exit;
}

// Define your API key (store securely in a real application)
define('API_KEY', '1234567890abcdef');

// Function to validate API key
function validateApiKey() {
    $headers = getallheaders();
    if (!isset($headers['X-API-KEY']) || $headers['X-API-KEY'] !== API_KEY) {
        echo json_encode([
            '$api_status' => 'error',
            '$message' => 'Forbidden: Invalid API Key'
        ]);
        http_response_code(403);
        exit;
    }
}

// Validate API Key before proceeding
validateApiKey();

// Function to extract all variable names from macros.conf
function getVariableNames() {
    $variables = [];
    
    if (file_exists('/etc/pf-include/macros.conf')) {
        $config = file_get_contents('/etc/pf-include/macros.conf');
        
        // Match all variable names
        preg_match_all('/^([a-zA-Z0-9_]+)=/m', $config, $matches);
        
        if (!empty($matches[1])) {
            $variables = array_unique($matches[1]); // Ensure unique variable names
        }
    }
    
    // Add $ sign before each variable
    $variables = array_map(fn($var) => "$$var", $variables);
    
    return $variables;
}

// Retrieve and return variable names
$response = [
    'api_status' => 'success',
    'message' => 'Variable names retrieved successfully',
    'variables' => getVariableNames()
];

echo json_encode($response, JSON_PRETTY_PRINT);

?>

File: Scripts/V2 21-06-2025/getMacros.py
#!/usr/local/bin/python3

import json
import re
import os
import sys

# Step 1: Output correct headers for CGI
print("Content-Type: application/json\n")

API_KEY = '1234567890abcdef'

# Step 2: API Key validation
api_key = os.environ.get('HTTP_X_API_KEY')
if api_key != API_KEY:
    print(json.dumps({
        'api_status': 'error',
        'message': 'Forbidden: Invalid API Key'
    }))
    sys.exit(0)

# Step 3: Get path info (e.g., /getMacros or /hello)
path = os.environ.get('PATH_INFO', '')

# Step 4: Route logic
if path == '/getMacros':
    # Original getMacros logic
    variables = []
    config_path = '/etc/pf-include/macros.conf'
    
    if os.path.exists(config_path):
        with open(config_path, 'r') as f:
            content = f.read()
            matches = re.findall(r'^([a-zA-Z0-9_]+)=', content, re.MULTILINE)
            if matches:
                variables = list(set(f"${var}" for var in matches))

    print(json.dumps({
        'api_status': 'success',
        'message': 'Variable names retrieved successfully',
        'variables': variables
    }, indent=4))

elif path == '/hello':
    # New /hello endpoint
    print(json.dumps({
        'api_status': 'success',
        'message': 'Hello, World!'
    }, indent=4))

else:
    # Catch-all for unknown routes
    print(json.dumps({
        'api_status': 'error',
        'message': f'Unknown endpoint: {path}'
    }, indent=4))
